examples/flutter_gallery/lib/demo/material/bottom_app_bar_demo.dart

```diff
@@ -0,0 +1,571 @@
+    for (int i = 0; i < _fabLocationConfigurations.length; i++) {
+      final _FabLocationConfiguration configuration = _fabLocationConfigurations[i];
+      options.add(
+          new DropdownMenuItem<int>(
+          child: new Text(configuration.name, overflow: TextOverflow.ellipsis),
+          value: i,
+        ),
+      );
+    }
+    return new SingleChildScrollView(
+      scrollDirection: Axis.horizontal,
+      primary: false,
+      child: DropdownButton<int>(items: options, onChanged: (int newIdx) { setState(() { fabLocationIndex = newIdx; }); }, value: fabLocationIndex),
+    );
+  }
+
+  Widget buildBabColorPicker() {
+    final List<Widget> colors = <Widget> [
+      const Text('Color:'),
+    ];
+    for (Color color in babColors) {
+      colors.add(
+        new Radio<Color>(
+          value: color,
+          groupValue: babColor,
+          onChanged: (Color color) { setState(() { babColor = color; }); },
+        )
+      );
+      colors.add(new Container(
+          decoration: new BoxDecoration(
+            color: color,
+            border: new Border.all(width:2.0, color: Colors.black),
+          ),
+          child: const SizedBox(width: 20.0, height: 20.0)
+      ));
+      colors.add(const Padding(padding: const EdgeInsets.only(left: 12.0)));
+    }
+    return new Row(
+      children: colors,
+      mainAxisAlignment: MainAxisAlignment.center,
+    );
+  }
+
+  static void _showSnackbar(BuildContext context) {
+    Scaffold.of(context).showSnackBar(const SnackBar(content: const Text(_explanatoryText)));
+  }
+}
+
+const String _explanatoryText =
+  "When the Scaffold's floating action button location changes, "
+  'the floating action button animates to its new position.'
+  'The BottomAppBar adapts its shape appropriately.';
+
+// Whether the Bottom App Bar's menu should keep icons away from the center or from the end of the screen.
+//
+// When the Floating Action Button is positioned at the end of the screen,
+// it would cover icons at the end of the screen, so the END_FAB mode tells 
+// the MyBottomAppBar to place icons away from the end.
+//
+// Similar logic applies to the CENTER_FAB mode.
+enum _BabMode {
+  END_FAB,
+  CENTER_FAB
+}
+
+// Pairs the Bottom App Bar's menu mode with a Floating Action Button Location.
+class _FabLocationConfiguration {
+  const _FabLocationConfiguration(this.name, this.babMode, this.fabLocation);
+
+  // The name of this configuration.
+  final String name;
+
+  // The _BabMode to place the menu in the bab with.
+  final _BabMode babMode;
+
+  // The location for the Floating Action Button.
+  final FloatingActionButtonLocation fabLocation;
+}
+
+// Map of names to the different shapes of Floating Action Button in this demo.
+class _FabShapeConfiguration {
+  const _FabShapeConfiguration(this.name, this.fab);
+
+  final String name;
+  final Widget fab;
+}
+
+// A bottom app bar with a menu inside it.
+class MyBottomAppBar extends StatelessWidget {
```

Comment:
For consistency, give this a _FooBottomAppBar name

Reply:
Done